### PR TITLE
chore: angular 4 upgrade

### PR DESCRIPTION
**Pull request by willyelm**
Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
Hi! Thnx for the PR. We can't merge it just yet as it would mean dropping support for Angular 2.x and we don't want to do it just yet. See #1337 for more details.

Having said this I'm going to keep this PR around so we know what / where to change when the update time comes.

**Comment by AndonyNS** [comment]
@pkozlowski-opensource  Have you guys considered moving the v4 migration to a different branch and make both available through npm so that people can use it both on angular 2 and 4? As the Angular folks do it.

File: tsconfig.json [hunk 1]
```diff
@@ -1,8 +1,9 @@
 {
   "compilerOptions": {
-    "target": "ES5",
-    "lib": ["es2015", "dom"],
+    "target": "es5",
+    "lib": ["es2016", "dom"],
```
Comment (pkozlowski-opensource) [inline tsconfig.json]: Why this change?

File: tsconfig.json [hunk 2]
```diff
@@ -1,8 +1,9 @@
...
     "module": "commonjs",
+    "moduleResolution": "node",
```
Comment (pkozlowski-opensource) [inline tsconfig.json]: Why this change?

File: demo/src/polyfills.ts [hunk 3]
```diff
@@ -1,14 +1,37 @@
...
 } else {
   // Development
-
   Error['stackTraceLimit'] = Infinity;
-
   require('zone.js/dist/long-stack-trace-zone');
+  require('zone.js/dist/proxy.js');
+  require('zone.js/dist/sync-test');
+  require('zone.js/dist/async-test');
+  require('zone.js/dist/fake-async-test');
```
Comment (pkozlowski-opensource) [inline demo/src/polyfills.ts]: I don't think we use those for unit tests so we shouldn't be having this change.

**Comment by pkozlowski-opensource** [comment]
@willyelm thnx for the PR - I've just started to review it but there are many changes in there and many of those are not related to the Angular 4 switch... Given this (and the fact that this PR doesn't merge cleanly any more) I'm going to pick up bits and pieces but won't be able to land the whole PR. I will leave comments here indicating which commits I'm brining in. Once again, thnx for the PR.

**Comment by willyelm** [comment]
Hi @pkozlowski-opensource,  Thanks! let me know if I can help.

**Comment by pkozlowski-opensource** [comment]
Merged WebPack config changes as 6616c77149239f54e2415ee52fab0744d36d55d8 - it was possible to land those now since not related to Angular 4 switch.
